src/linting/checks.jl

```diff
@@ -860,3 +860,132 @@ function check_const(x::EXPR)
         end
     end
 end
+
+function check_unused_binding(b::Binding, scope::Scope)
+    if headof(scope.expr) !== :struct && headof(scope.expr) !== :tuple && !all_underscore(valof(b.name))
+        if (isempty(b.refs) || length(b.refs) == 1 && b.refs[1] == b.name) && !is_sig_arg(b.name) && !is_overwritten_in_loop(b.name) && !is_overwritten_subsequently(b, scope)
+            seterror!(b.name, UnusedBinding)    
+        end
+    end
+end
+
+all_underscore(s) = false
+all_underscore(s::String) = all(==(0x5f), codeunits(s))
+
+function is_sig_arg(x)
+    is_in_fexpr(x, CSTParser.iscall)
+end
+
+function is_overwritten_in_loop(x)
+    # Cuts out false positives for check_unused_binding - the linear nature of our 
+    # semantic passes mean a variable declared at the end of a loop's block but used at 
+    # the start won't appear to be referenced.
```

Comment:
Hm, but isn't that in line with how Julia's loops work?
```
julia> for i in 1:3
         if i > 2
           println(x)
         end
         x = i
       end
ERROR: UndefVarError: x not defined
```
So we only need to check if the variable being assigned is used in a enclosing scope, as you mention below?

Reply:
Hmm, that i did not know!

Reply:
Doesn't matter whether it's in a top-level scope or not, afaict:
```
julia> function f()
         for i in 1:3
           if i > 2
             println(x)
           end
           x=i
         end
       end
f (generic function with 1 method)

julia> f()
ERROR: UndefVarError: x not defined
```

Reply:
I meant for the StaticLint checks